src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/UserTimeZoneService.cs

```diff
@@ -5,83 +5,97 @@
+    {
+        var currentTimeZoneId = await GetCurrentUserTimeZoneIdAsync();
+
+        if (string.IsNullOrEmpty(currentTimeZoneId))
         {
-            _clock = clock;
-            _distributedCache = distributedCache;
-            _httpContextAccessor = httpContextAccessor;
-            _userManager = userManager;
+            return null;
         }
 
-        public async Task<ITimeZone> GetUserTimeZoneAsync()
-        {
-            var currentTimeZoneId = await GetCurrentUserTimeZoneIdAsync();
-            if (string.IsNullOrEmpty(currentTimeZoneId))
-            {
-                return null;
-            }
+        return _clock.GetTimeZone(currentTimeZoneId);
+    }
 
-            return _clock.GetTimeZone(currentTimeZoneId);
-        }
+    public async Task UpdateUserTimeZoneAsync(IUser user)
+    {
+        var userName = user?.UserName;
 
-        public Task UpdateUserTimeZoneAsync(User user)
+        if (!string.IsNullOrEmpty(userName))
         {
-            var userName = user?.UserName;
+            await _distributedCache.RemoveAsync(GetCacheKey(userName));
+        }
 
-            if (!string.IsNullOrEmpty(userName))
-            {
-                return _distributedCache.RemoveAsync(GetCacheKey(userName));
-            }
+        return;
+    }
 
-            return Task.CompletedTask;
-        }
+    public async Task<string> GetCurrentUserTimeZoneIdAsync()
+    {
+        var userName = _httpContextAccessor.HttpContext.User?.Identity?.Name;
 
-        public async Task<string> GetCurrentUserTimeZoneIdAsync()
+        if (string.IsNullOrEmpty(userName))
         {
-            var userName = _httpContextAccessor.HttpContext.User?.Identity?.Name;
+            return null;
+        }
 
-            if (string.IsNullOrEmpty(userName))
-            {
-                return null;
-            }
+        var key = GetCacheKey(userName);
+        var timeZoneId = await _distributedCache.GetStringAsync(key);
 
-            var key = GetCacheKey(userName);
-            var timeZoneId = await _distributedCache.GetStringAsync(key);
+        if (string.IsNullOrEmpty(timeZoneId))
+        {
+            var user = await _userManager.FindByNameAsync(userName) as User;
+            timeZoneId = user.As<UserTimeZone>()?.TimeZoneId;
 
             if (string.IsNullOrEmpty(timeZoneId))
             {
-                var user = await _userManager.FindByNameAsync(userName) as User;
-                timeZoneId = user.As<UserTimeZone>()?.TimeZoneId;
-
-                if (!string.IsNullOrEmpty(timeZoneId))
-                {
-                    await _distributedCache.SetStringAsync(key, timeZoneId, new DistributedCacheEntryOptions { SlidingExpiration = _slidingExpiration });
-                }
+                var siteSettings = await _siteService.GetSiteSettingsAsync();
+                timeZoneId = string.IsNullOrEmpty(siteSettings.TimeZoneId)
+                    ? _clock.GetSystemTimeZone().TimeZoneId
```

Comment:
Isn't this doing the job of the TimeZone selector? I would just have stored an empty value in the cache to let the selector know there is nothing there and fallback to its own logic.

Reply:
Yes, you are right because the TimeZoneSelector uses GetUserTimeZoneAsync which has the fallback logic already. So allowing to store a NULL value in the cache should fix the issue.